fix another case where we might return free TypeVar

vtjnash opened the pull request:
Still one error caused here (this intersection is not satisfiable except by `Union{}`)
```julia
julia> T=Tuple{Type{Q}, Q, Ref{Q}} where Q<:Ref;

julia> S=Tuple{Type{S}, Union{Ref{S}, Ref{R}}, R} where R where S;

julia> typeintersect(S,T)
Tuple{Type{Q}, Q, R} where {Q<:(Ref), R<:Ref{Q}}
                         
julia> typeintersect(T,S)
StackOverflowError
```

Original PR text:

---

We had an environment here that looked like while computing the upper bound for J{S} where S:

where S=T
where T
where I{T}
where J{S} where S

Then we started handling those, and filling in the values:

First replacing S with T, which creates a `res` of `J{T}`

where T
where I{T}
where J{S} where S

Then we handled T, which is also going to set `wrap=0`, so our result for `J{T}` will not be made into `J{T} where T`.

where I{T} (wrap 0)
where J{S} where S

Here we then had finished handling all the dependencies for J{S} where S, which resulted in an upper bound assignment of J{T}

where I{T}
where J{T}

Next, we handle I{T}, though it is now unused, so while we will make `I{T} where T` (via innervars) here for it, this goes unuesd.

And finally, we had our resulting clause:

where J{T}

But it is missing the `where T`, since `I` (from lhs) was discarded.

Thus we need to add that back now, when handling some innervars, if we see our term got duplicated to a higher part of the bounds before reaching this handling for its placement movement.

vtjnash committed:
fix another case where we might return free TypeVar

We had an environment here that looked like while computing the upper bound for J{S} where S:

where S=T
where T
where I{T}
where J{S} where S

Then we started handling those, and filling in the values:

First replacing S with T, which creates a `res` of `J{T}`

where T
where I{T}
where J{S} where S

Then we handled T, which is also going to set `wrap=0`, so our result
for `J{T}` will not be made into `J{T} where T`.

where I{T} (wrap 0)
where J{S} where S

Here we then had finished handling all the dependencies for J{S} where
S, which resulted in an upper bound assignment of J{T}

where I{T}
where J{T}

Next, we handle I{T}, though it is now unused, so while we will make `I{T}
where T` (via innervars) here for it, this goes unuesd.

And finally, we had our resulting clause:

where J{T}

But it is missing the `where T`, since `I` (from lhs) was discarded.

Thus we need to add that back now, when handling some innervars, if we
see our term got duplicated to a higher part of the bounds before
reaching this handling for its placement movement.

N5N3 commented:
Adding too many inner-vars might not be a good idea.
As inner-var is not in stenv and would be ignored during subtyping.
But the wrap change seems reasonable. We'd better assign the inner-var with the outmost in-env-var rather than wrap it many times.

vtjnash commented:
Upon review, I think we should attempt to go ahead with this now. It turns out the intersection in the tests that stackoverflows, would have instead become incorrect if it was simplified:
```
julia> T = Tuple{Type{Q}, Q, Ref{Q}} where Q<:Ref;

julia> S = Tuple{Type{S}, Ref{S}, R} where R where S;

julia> typeintersect(S,T)
Union{}

julia> typeintersect(T,S)
Union{}

julia> I = Tuple{Type{Ref}, Ref{Ref}, Ref{Ref}}; I<:T, I<:S
(true, true)
```
In particular, I believe `I = Tuple{Type{Ref}, Ref{Ref}, Ref{Ref}}` may be the one-true-intersect for these types. So I think that, on net, this PR seems to likely be an improvement.

N5N3 commented:
@nanosoldier `runtests()`

N5N3 commented:
> In particular, I believe I = Tuple{Type{Ref}, Ref{Ref}, Ref{Ref}}

I think there have been some discussion for this pattern.
As you pointed out in https://github.com/JuliaLang/julia/pull/38245#discussion_r516031979, perhaps we should return `Tuple{Type{T}, Ref{T}, Ref{T}} where {T<:Ref}` in this case.

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48228#issuecomment-1472988061) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/d48110d_vs_c1d1bde/report.html).

vtjnash commented:
I see that did quite poorly

N5N3 committed:
Fix the stackoverflow

Should also fix some stackoverflow in PkgEval

vtjnash commented:
@nanosoldier `runtests(["PiecewiseLinearOpt", "Sudoku", "ConstraintModels", "NumericalAlgorithms", "ParametricLP", "QAOA", "SpeedMapping", "MultiObjectiveAlgorithms", "PowerModelsAnnex", "SymbolicWedderburn", "Nonconvex", "PowerModelsACDC", "QuantumInformation", "SDDP", "Clp", "CacheVariables", "NonconvexSearch", "PowerModelsRestoration", "NonconvexJuniper", "ModelPredictiveControl", "HorseML", "MultiscaleGraphSignalTransforms", "JuDGE", "BDisposal", "TrajectoryOptimization", "PoGO", "PowerModelsProtection", "ToQUBO", "UnitJuMP", "SumOfSquares", "ReliabilityOptimization", "Effects", "MathOptSetDistances", "UnitCommitment", "SideKicks", "NonconvexPavito", "QuantumTomography", "DigitalComm", "EBayes", "StandaloneIpopt", "BlockDecomposition", "vOptGeneric", "MicrobiomeAnalysis", "ECOS", "MixedModels", "DelaySSAToolkit", "NonconvexNOMAD", "SetProg", "InteractBulma", "BellDiagonalQudits", "ExpressionTreeForge", "RealTimeScheduling", "SNOW", "QHull", "PointEstimateMethod", "PetroleumModels", "TrajectoryGamesBase", "NetDecOPF", "AnyMOD", "REoptLite", "MixedModelsSim", "ITensorVisualizationBase", "FlexPlan", "ProjectAssigner", "MultistartOptimization", "Unfold", "OpSel", "CorrelatedEquilibria", "PhyloNetworks", "IndependentHypothesisWeighting", "HetaSimulator", "NoncommutativeGraphs", "NonconvexPercival", "OSQP", "TrajGWAS", "MixedModelsPermutations", "RegularizedProblems", "BEASTDataPrep", "CompressiveLearning", "HybridSystems", "Powersense", "PhyloPlots", "FrankWolfe", "GasPowerModels", "StaticStrings", "RDMREopt", "Juniper", "BayesianOptimization", "CalibrationAnalysis", "RandomFeatures", "LightGraphsMatching", "CompressedSensing", "DAQP", "Bonobo", "NeRCA", "StoppingInterface", "PortfolioAnalytics", "SCIP", "PhyloDiamond", "RoundAndSwap", "SlimOptim", "Fri", "DecisionProgramming", "StructJuMP", "MathematicalSystems", "ProjectManagement", "Allocations", "MixedModelsSerialization", "EfficientGlobalOptimization", "LightGraphsFlows", "SCS", "RecordedArrays", "PowerModelsADA", "OrdinalMultinomialModels", "SecretSanta", "ParametricOptInterface", "PolyhedralRelaxations", "QUnfold", "LatinSquares", "PartiallySeparableNLPModels", "HiGHS", "DWaveNeal", "SchattenNorms", "RigorousInvariantMeasures", "Boscia", "SqraCore", "OptimizationNOMAD", "VoronoiGraph", "SparseVariables", "CutPruners", "ONNXNaiveNASflux", "RobustFactorizations", "QuantumCircuitOpt", "SymbolicRegression", "DataEnvelopmentAnalysis", "PowerWaterModels", "FinEtools", "PowerPlots", "LinRegOutliers", "BasisMatrices", "MultiwayNumberPartitioning", "SteadyStateDiffEq", "ThresholdStability", "ParameterEstimocean", "LPVSpectral", "CALiPPSO", "CSVReader", "BundleMethod", "LRSLib", "LovaszTheta", "BlackBoxOptimizationBenchmarking", "Cloudy", "UnfoldSim", "SeparableOptimization", "GameTheory", "UlamMethod", "DirectTrajectoryOptimization", "Subspaces", "PolynomialAmoebas", "GLPK", "Alpine", "MetidaCu", "BioGraph", "ClosedLoopReachability", "DifferentiableTrajectoryOptimization", "ExaTron", "CalibrationTests", "RegressionDiscontinuity", "Bloqade", "Knet", "ProximalAlgorithms", "LayeredLayouts", "RobustModels", "MQLib", "VectorizationBase", "GraphsMatching", "OptimalPortfolios", "GuessworkQuantumSideInfo", "C3D", "UnfoldMakie", "Knapsacks", "Ipopt", "ParameterJuMP", "StatsKit", "AnovaMixedModels", "OPFSampler", "Phylo", "MINLPTests", "BijectiveHilbert", "ParametrisedConvexApproximators", "Octavian", "Pavito", "JLBoost", "ConstrainedShortestPaths", "COPT", "ITensorUnicodePlots", "CSDP", "OrdinalGWAS", "Anneal", "PowerModelsSecurityConstrained", "MaximinOPF", "Graphene", "DataGraphs", "BenchmarkingEconomicEfficiency", "POMDPStressTesting", "ComplexOptInterface", "StatsDiscretizations", "SimplePosetAlgorithms", "Vecchia", "Anatta", "NonconvexIpopt", "Seleroute", "GraphicalModelLearning", "Plasmo", "PowerModelsStability", "GasModels", "WaterModels", "BellPolytopes", "QuartetNetworkGoodnessFit", "NaiveNASlib", "PowerModelsWildfire", "Tapestree", "Dualization", "InfrastructureModels", "MixedModelsMakie", "PowerModelsITD", "YasolSolver", "JuMP", "CBLS", "ProxSDP", "MultiJuMP", "JMcDM", "MixedModelsExtras", "PowerAnalytics", "LinearFractional", "KrigingModel", "GraphSignals", "CalibrationErrorsDistributions", "MathOptSymbolicAD", "REopt", "CVaRRiskParity", "GraphsFlows", "BranchAndBound", "ReachabilityAnalysis", "DiscreteFunctions", "CompEcon", "PWF", "RheaMetacycParser", "VoronoiFVMDiffEq", "Markets", "MIPVerify", "RangeEnclosures", "NLPModelsIpopt", "SDPA", "Cbc", "NextGP", "ConstraintSolver", "OptimalTransport", "ClimaCore"])`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48228#issuecomment-1483909767) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f18fe28_vs_703b3f8/report.html).

N5N3 commented:
Looks like most of the problem get fixed.

vtjnash commented:
Yes, looks like we are down to 2 new internal failures, and 3 fixed subtyping results (causing ambiguity tests to fail)

vtjnash committed:
fixup algorithm

Had an issue where (S<:T<:S with innervar N) would add the (where N) to
each S instead of the whole expression, resulting to creation of
(Val{. where N}) instead of (Val{.} where N) later.

vtjnash commented:
I think we are starting to converge here towards something usable. It looks like one of N5N3's tests successfully demonstrated why the innervars update was needed (which I had previously commented out, in the absence of a test case for it). The failing package test inference looks something like:

```julia
(gdb) p jl_(a)
Tuple{typeof(Phylo.API._getnode), T, N} where T<:(Phylo.AbstractTree{Phylo.OneTree, RT, NL, N, B} where B<:Phylo.AbstractBranch{RT, NL}) where N where NL where RT

(gdb) p (void*) jl_call2((void*) jl_get_global(jl_base_module, (void*)jl_symbol("show")), (void*) jl_get_global(jl_core_module, (void*)jl_symbol("stdout")), types)
Tuple{typeof(Phylo.API._getnode), Union{Union{LinkTree{RT3, NL, N, B} where {RT2<:Union{RT, RT1}, N<:Union{LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{RT2, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT2, NL}}, LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}}, B<:Union{LinkBranch{RT, NL}, LinkBranch{RT2, NL}, LinkBranch{RT1, NL}}, RT3<:Union{RT, RT2, RT1}}, LinkTree{var"#s63", NL, N, B} where {var"#s63"<:Union{RT, RT1}, N<:Union{LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s63", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s63", NL}}, LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}}, B<:Union{LinkBranch{RT, NL}, LinkBranch{var"#s63", NL}, LinkBranch{RT1, NL}}}} where {RT<:Phylo.Rooted, RT1<:Phylo.Rooted}, Union{LinkTree{RT2, NL, N, B} where {RT2<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{RT2, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT2, NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{RT2, NL}}}, LinkTree{var"#s63", NL, N, B} where {var"#s63"<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s63", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s63", NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{var"#s63", NL}}}} where {RT<:Phylo.Rooted, RT1<:Phylo.Rooted}, Union{LinkTree{var"#s65", NL, N, B} where {var"#s65"<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s65", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s65", NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{var"#s65", NL}}}, LinkTree{var"#s63", NL, N, B} where {var"#s63"<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s63", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s63", NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{var"#s63", NL}}}} where {RT<:Phylo.Rooted, RT1<:Phylo.Rooted}} where NL<:Missing, Missing}

caused by inference of
branchroute(Phylo.LinkTree{Phylo.OneRoot, String, Phylo.LinkNode{Phylo.OneRoot, String, Base.Dict{String, Any}, Phylo.LinkBranch{Phylo.OneRoot, String, Base.Dict{String, Any}, Float64}}, Phylo.LinkBranch{Phylo.OneRoot, String, Base.Dict{String, Any}, Float64}, Base.Dict{String, Any}}, String, String)
from branchroute(T, N, N) where {RT, N, T<:(Phylo.AbstractTree{Phylo.OneTree, RT, NL, N, B} where B<:Phylo.AbstractBranch{RT, NL} where N<:Phylo.AbstractNode{RT, NL} where NL)}
```
where we eventually get stuck (recurse infinitely) on something like the intersection of these values:
```
RT<:Union{Phylo.Rooted, RT<:Phylo.Rooted}
Union{Phylo.Rooted, RT}
```
(Aside: it also makes me feel very uncomfortable that inference is producing that mess in the first place also. This might end up being worked around by https://github.com/JuliaLang/julia/pull/49167, or something like that PR.)

N5N3 commented on src/subtype.c:
```diff
@@ -2847,10 +2850,23 @@ static jl_value_t *finish_unionall(jl_value_t *res JL_MAYBE_UNROOTED, jl_varbind
+            // the `btemp->prev` walk is only giving a sort of post-order guarantee (since we are
+            // iterating 2 trees at once), so once we set `wrap`, there might remain other branches
+            // of the type walk that now may have incomplete bounds: finish those now too
+            jl_varbinding_t *btemp = e->vars;
```
Perhaps we still need to make sure there no duplicated innervar in the env (just like the `jl_varbinding_t *wrap = NULL;` trick above)?

And the TODO seems get resolved?

N5N3 replied:
And perhaps we'd better not wrap `res` if we update the `innervars`?

vtjnash replied:
From my original analysis, I didn't expect that would be possible. There are only 2 sides to the type tree, so either wrapping to the left or wrapping to the right should be needed, but not more than that. With an attempt of asserting for this case, I don't think it is hit either, but the code rearrangement probably makes sense anyways to be more straightforward. I believe a variable would have to switch sides multiple times by combinations to get tricky here, and it doesn't seem like that happens in existing code.

vtjnash commented:
Further call stack analysis, which can be simply by plugging this into `Core.Compiler.return_type`. This package is quite designed to be quite uninferable apparently, so Cthulhu has absolutely no idea what is going on with inference, sometimes even annotating calls with multiple conflicting remarks and contains some messy Unions almost immediately from the user.
```
hasinbound(Type{SimpleTraits.Not{Phylo.API.MatchNodeType{T, NL}}}, T, NL) where {NL, T<:(Phylo.AbstractTree{Phylo.OneTree, var"#s63", NL, N, B} where B<:Phylo.AbstractBranch{var"#s63", NL} where N<:Phylo.AbstractNode{var"#s63", NL} where var"#s63"<:Phylo.Rooted)}
hasinbound(T, NL) where {NL, T<:(Phylo.AbstractTree{Phylo.OneTree, var"#s65", NL, N, B} where B<:Phylo.AbstractBranch{var"#s65", NL} where N<:Phylo.AbstractNode{var"#s65", NL} where var"#s65"<:Phylo.Rooted)}
_treehistory(Type{Phylo.API.PreferBranchObjects{T}}, T, N1) where {N1, NL, N, B, T<:(Phylo.AbstractTree{Phylo.OneTree, var"#s63", NL, N, B} where var"#s63"<:Phylo.Rooted)}
_treehistory(T, N1) where {N1, NL, N, B, T<:(Phylo.AbstractTree{Phylo.OneTree, var"#s65", NL, N, B} where var"#s65"<:Phylo.Rooted)}
branchroute(Type{Phylo.API.MatchNodeType{T, N}}, T, N, N) where {RT, N, T<:(Phylo.AbstractTree{Phylo.OneTree, RT, NL, N, B} where B<:Phylo.AbstractBranch{RT, NL} where N<:Phylo.AbstractNode{RT, NL} where NL)}
branchroute(T, N, N) where {RT, N, T<:(Phylo.AbstractTree{Phylo.OneTree, RT, NL, N, B} where B<:Phylo.AbstractBranch{RT, NL} where N<:Phylo.AbstractNode{RT, NL} where NL)}
branchroute(Type{SimpleTraits.Not{Phylo.API.MatchNodeType{T, N}}}, T, N, N) where {RT, N, T<:(Phylo.AbstractTree{Phylo.OneTree, RT, NL, N, B} where B<:Phylo.AbstractBranch{RT, NL} where N<:Phylo.AbstractNode{RT, NL} where NL)}
branchroute(T, N, N) where {RT, N, T<:(Phylo.AbstractTree{Phylo.OneTree, RT, NL, N, B} where B<:Phylo.AbstractBranch{RT, NL} where N<:Phylo.AbstractNode{RT, NL} where NL)}

specialized as

Tuple{typeof(hasinbound), Type{SimpleTraits.Not{Phylo.API.MatchNodeType{T, NL}}}, Union{Union{LinkTree{RT3, NL, N, B} where {RT2<:Union{RT, RT1}, N<:Union{LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{RT2, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT2, NL}}, LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}}, B<:Union{LinkBranch{RT, NL}, LinkBranch{RT2, NL}, LinkBranch{RT1, NL}}, RT3<:Union{RT, RT2, RT1}}, LinkTree{var"#s63", NL, N, B} where {var"#s63"<:Union{RT, RT1}, N<:Union{LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s63", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s63", NL}}, LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}}, B<:Union{LinkBranch{RT, NL}, LinkBranch{var"#s63", NL}, LinkBranch{RT1, NL}}}} where {RT<:Phylo.Rooted, RT1<:Phylo.Rooted}, Union{LinkTree{RT2, NL, N, B} where {RT2<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{RT2, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT2, NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{RT2, NL}}}, LinkTree{var"#s63", NL, N, B} where {var"#s63"<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s63", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s63", NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{var"#s63", NL}}}} where {RT<:Phylo.Rooted, RT1<:Phylo.Rooted}, Union{LinkTree{var"#s65", NL, N, B} where {var"#s65"<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s65", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s65", NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{var"#s65", NL}}}, LinkTree{var"#s63", NL, N, B} where {var"#s63"<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s63", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s63", NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{var"#s63", NL}}}} where {RT<:Phylo.Rooted, RT1<:Phylo.Rooted}}, NL} where {NL<:Missing, T<:(AbstractTree{OneTree, var"#s63", NL, N, B} where {var"#s63"<:Phylo.Rooted, N<:Phylo.AbstractNode{var"#s63", NL}, B<:Phylo.AbstractBranch{var"#s63", NL}})}

Tuple{typeof(hasinbound), T, NL} where {NL<:Missing, T<:Union{LinkTree{RT3, NL, N, B} where {RT<:Phylo.Rooted, RT1<:Phylo.Rooted, RT2<:Union{RT, RT1}, N<:Union{LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{RT2, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT2, NL}}, LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}}, B<:Union{LinkBranch{RT, NL}, LinkBranch{RT2, NL}, LinkBranch{RT1, NL}}, RT3<:Union{RT, RT2, RT1}}, Union{LinkTree{RT2, NL, N, B} where {RT2<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{RT2, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT2, NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{RT2, NL}}}, LinkTree{var"#s65", NL, N, B} where {var"#s65"<:Union{Phylo.Rooted, RT1, RT}, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}, LinkNode{var"#s65", NL, Data, B} where {Data, B<:Phylo.AbstractBranch{var"#s65", NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}, LinkBranch{var"#s65", NL}}}} where {RT<:Phylo.Rooted, RT1<:Phylo.Rooted}}}


Tuple{typeof(Phylo._treehistory), Type{Phylo.API.PreferBranchObjects{T}}, T, Missing} where {NL, RT<:Phylo.Rooted, RT1<:Union{Phylo.Rooted, RT}, RT, RT2<:Union{Phylo.Rooted, RT}, RT, N<:Union{LinkNode{RT2, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT2, NL}}, LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}}, B<:Union{LinkBranch{RT2, NL}, LinkBranch{RT1, NL}, LinkBranch{RT, NL}}, T<:Union{LinkTree{RT1, NL, N, B} where RT1<:Union{RT2, RT}, LinkTree{RT, NL, N, B}}}


Tuple{typeof(Phylo._treehistory), T, Missing} where {NL, RT<:Phylo.Rooted, RT1<:Phylo.Rooted, N<:Union{LinkNode{RT1, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT1, NL}}, LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}}, B<:Union{LinkBranch{RT1, NL}, LinkBranch{RT, NL}}, T<:Union{LinkTree{RT1, NL, N, B}, LinkTree{RT, NL, N, B}}}


Tuple{typeof(branchroute), Type{Phylo.API.MatchNodeType{T, N}}, Union{LinkTree{RT, NL, N, B} where {NL, N<:(LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}), B<:(LinkBranch{RT, NL})}, LinkTree{RT, NL, N, B} where {NL, N<:(LinkNode{RT, NL, Data, B} where {Data, B<:Phylo.AbstractBranch{RT, NL}}), B<:(LinkBranch{RT, NL})}}, Missing, Missing} where {RT, N, T<:(AbstractTree{OneTree, RT, NL, N, B} where {NL, N<:Phylo.AbstractNode{RT, NL}, B<:Phylo.AbstractBranch{RT, NL}})}

Tuple{typeof(branchroute), LinkTree, Missing, Missing}

Tuple{typeof(branchroute), Type{SimpleTraits.Not{Phylo.API.MatchNodeType{RootedTree, String}}}, RootedTree, String, String}

Tuple{typeof(branchroute), RootedTree, String, String}
```

vtjnash committed:
some algorithm layout cleanup

I think this might be NFC, since I don't think this can simultaneously
affect the left and right envs, but it looks better this way.

vtjnash committed:
fixup! some algorithm layout cleanup

vtjnash committed:
fixup! fixup! some algorithm layout cleanup

Correct test for some other improvement also now on master.

vtjnash commented:
@nanosoldier `runtests(["PiecewiseLinearOpt", "Sudoku", "ConstraintModels", "NumericalAlgorithms", "ParametricLP", "QAOA", "SpeedMapping", "MultiObjectiveAlgorithms", "PowerModelsAnnex", "SymbolicWedderburn", "Nonconvex", "PowerModelsACDC", "QuantumInformation", "SDDP", "Clp", "CacheVariables", "NonconvexSearch", "PowerModelsRestoration", "NonconvexJuniper", "ModelPredictiveControl", "HorseML", "MultiscaleGraphSignalTransforms", "JuDGE", "BDisposal", "TrajectoryOptimization", "PoGO", "PowerModelsProtection", "ToQUBO", "UnitJuMP", "SumOfSquares", "ReliabilityOptimization", "Effects", "MathOptSetDistances", "UnitCommitment", "SideKicks", "NonconvexPavito", "QuantumTomography", "DigitalComm", "EBayes", "StandaloneIpopt", "BlockDecomposition", "vOptGeneric", "MicrobiomeAnalysis", "ECOS", "MixedModels", "DelaySSAToolkit", "NonconvexNOMAD", "SetProg", "InteractBulma", "BellDiagonalQudits", "ExpressionTreeForge", "RealTimeScheduling", "SNOW", "QHull", "PointEstimateMethod", "PetroleumModels", "TrajectoryGamesBase", "NetDecOPF", "AnyMOD", "REoptLite", "MixedModelsSim", "ITensorVisualizationBase", "FlexPlan", "ProjectAssigner", "MultistartOptimization", "Unfold", "OpSel", "CorrelatedEquilibria", "PhyloNetworks", "IndependentHypothesisWeighting", "HetaSimulator", "NoncommutativeGraphs", "NonconvexPercival", "OSQP", "TrajGWAS", "MixedModelsPermutations", "RegularizedProblems", "BEASTDataPrep", "CompressiveLearning", "HybridSystems", "Powersense", "PhyloPlots", "FrankWolfe", "GasPowerModels", "StaticStrings", "RDMREopt", "Juniper", "BayesianOptimization", "CalibrationAnalysis", "RandomFeatures", "LightGraphsMatching", "CompressedSensing", "DAQP", "Bonobo", "NeRCA", "StoppingInterface", "PortfolioAnalytics", "SCIP", "PhyloDiamond", "RoundAndSwap", "SlimOptim", "Fri", "DecisionProgramming", "StructJuMP", "MathematicalSystems", "ProjectManagement", "Allocations", "MixedModelsSerialization", "EfficientGlobalOptimization", "LightGraphsFlows", "SCS", "RecordedArrays", "PowerModelsADA", "OrdinalMultinomialModels", "SecretSanta", "ParametricOptInterface", "PolyhedralRelaxations", "QUnfold", "LatinSquares", "PartiallySeparableNLPModels", "HiGHS", "DWaveNeal", "SchattenNorms", "RigorousInvariantMeasures", "Boscia", "SqraCore", "OptimizationNOMAD", "VoronoiGraph", "SparseVariables", "CutPruners", "ONNXNaiveNASflux", "RobustFactorizations", "QuantumCircuitOpt", "SymbolicRegression", "DataEnvelopmentAnalysis", "PowerWaterModels", "FinEtools", "PowerPlots", "LinRegOutliers", "BasisMatrices", "MultiwayNumberPartitioning", "SteadyStateDiffEq", "ThresholdStability", "ParameterEstimocean", "LPVSpectral", "CALiPPSO", "CSVReader", "BundleMethod", "LRSLib", "LovaszTheta", "BlackBoxOptimizationBenchmarking", "Cloudy", "UnfoldSim", "SeparableOptimization", "GameTheory", "UlamMethod", "DirectTrajectoryOptimization", "Subspaces", "PolynomialAmoebas", "GLPK", "Alpine", "MetidaCu", "BioGraph", "ClosedLoopReachability", "DifferentiableTrajectoryOptimization", "ExaTron", "CalibrationTests", "RegressionDiscontinuity", "Bloqade", "Knet", "ProximalAlgorithms", "LayeredLayouts", "RobustModels", "MQLib", "VectorizationBase", "GraphsMatching", "OptimalPortfolios", "GuessworkQuantumSideInfo", "C3D", "UnfoldMakie", "Knapsacks", "Ipopt", "ParameterJuMP", "StatsKit", "AnovaMixedModels", "OPFSampler", "Phylo", "MINLPTests", "BijectiveHilbert", "ParametrisedConvexApproximators", "Octavian", "Pavito", "JLBoost", "ConstrainedShortestPaths", "COPT", "ITensorUnicodePlots", "CSDP", "OrdinalGWAS", "Anneal", "PowerModelsSecurityConstrained", "MaximinOPF", "Graphene", "DataGraphs", "BenchmarkingEconomicEfficiency", "POMDPStressTesting", "ComplexOptInterface", "StatsDiscretizations", "SimplePosetAlgorithms", "Vecchia", "Anatta", "NonconvexIpopt", "Seleroute", "GraphicalModelLearning", "Plasmo", "PowerModelsStability", "GasModels", "WaterModels", "BellPolytopes", "QuartetNetworkGoodnessFit", "NaiveNASlib", "PowerModelsWildfire", "Tapestree", "Dualization", "InfrastructureModels", "MixedModelsMakie", "PowerModelsITD", "YasolSolver", "JuMP", "CBLS", "ProxSDP", "MultiJuMP", "JMcDM", "MixedModelsExtras", "PowerAnalytics", "LinearFractional", "KrigingModel", "GraphSignals", "CalibrationErrorsDistributions", "MathOptSymbolicAD", "REopt", "CVaRRiskParity", "GraphsFlows", "BranchAndBound", "ReachabilityAnalysis", "DiscreteFunctions", "CompEcon", "PWF", "RheaMetacycParser", "VoronoiFVMDiffEq", "Markets", "MIPVerify", "RangeEnclosures", "NLPModelsIpopt", "SDPA", "Cbc", "NextGP", "ConstraintSolver", "OptimalTransport", "ClimaCore"])`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48228#issuecomment-1500606294) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/67e2297_vs_14afdd2/report.html).

vtjnash commented:
@N5N3 Phylo is a deterministic crash caused by omit_bad_union. In particular, the bound of one of the vars is being set to NULL here:
```
2833            if (jl_has_typevar(btemp->ub, vb->var)) {
2834                if (vb->ub == (jl_value_t*)btemp->var) {
2835 >>>                btemp->ub = omit_bad_union(btemp->ub, vb->var); <<<
2836                    if (btemp->ub == NULL) {
2837                        JL_GC_POP();
2838                        return jl_bottom_type;
2839                    }
2840                }
```
But I don't think the code expects these to ever be NULL, so a later call to `omit_bad_union` then segfaults. Did you intend this state to be valid, or should we reset those to something legal (bottom) immediately:
```diff
diff --git a/src/subtype.c b/src/subtype.c
index 7373db4e01..38b07db8d6 100644
--- a/src/subtype.c
+++ b/src/subtype.c
@@ -2834,6 +2834,7 @@ static jl_value_t *finish_unionall(jl_value_t *res JL_MAYBE_UNROOTED, jl_varbind
             if (vb->ub == (jl_value_t*)btemp->var) {
                 btemp->ub = omit_bad_union(btemp->ub, vb->var);
                 if (btemp->ub == NULL) {
+                    btemp->ub = jl_bottom_type;
                     JL_GC_POP();
                     return jl_bottom_type;
                 }
@@ -2998,7 +2999,7 @@ static jl_value_t *intersect_unionall_(jl_value_t *t, jl_unionall_t *u, jl_stenv
             vb->ub = omit_bad_union(vb->ub, u->var);
             JL_GC_POP();
             if (vb->ub == NULL)
-                res = jl_bottom_type;
+                res = vb->ub = jl_bottom_type;
         }
     }
     if (res != jl_bottom_type)
```

N5N3 committed:
let `omit_bad_union` return `Union{}` instead of `NULL`

N5N3 commented:
I thought setting `res = `Union{}` is enough to avoid future accessing. But apparently, we get fooled here.
[301f57d](https://github.com/JuliaLang/julia/pull/48228/commits/301f57dadb207d03b0e80104de5e3202f48c3504) changes some test 
```julia
julia> typeintersect(Tuple{Type{T}, T, Ref{T}} where T,
                         Tuple{Type{S}, Ref{S}, S} where S)
Tuple{Type{Union{}}, Union{}, Union{}}
```
But I think there's no need to change the test as #49111 would "fix" it.

N5N3 committed:
Merge branch 'master' into jn/47476more3
